// relay-compiler/lib/RelayCodeGenerator.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */
'use strict';

function generate(node) {
  switch (node.kind) {
    case 'Fragment':
      return require("./ReaderCodeGenerator").generate(node);

    case 'Request':
      return {
        kind: 'Request',
        fragment: require("./ReaderCodeGenerator").generate(node.fragment),
        operation: require("./NormalizationCodeGenerator").generate(node.root),
        params: {
          operationKind: node.root.operation,
          name: node.name,
          id: node.id,
          text: node.text,
          metadata: node.metadata
        }
      };

    case 'SplitOperation':
      return require("./NormalizationCodeGenerator").generate(node);
  }

  throw require("./RelayCompilerError").createCompilerError("RelayCodeGenerator: Unknown AST kind '".concat(node.kind, "'."), [node.loc]);
}

module.exports = {
  generate: generate
};
