// relay-compiler/lib/RelayFlowTypeTransformers.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *  strict-local
 * @format
 */
'use strict';

function getInputObjectTypeIdentifier(type) {
  return type.name;
}

function transformScalarType(type, state, objectProps) {
  if (type instanceof require("graphql").GraphQLNonNull) {
    return transformNonNullableScalarType(type.ofType, state, objectProps);
  } else {
    return require("@babel/types").nullableTypeAnnotation(transformNonNullableScalarType(type, state, objectProps));
  }
}

function transformNonNullableScalarType(type, state, objectProps) {
  if (type instanceof require("graphql").GraphQLList) {
    return require("./RelayFlowBabelFactories").readOnlyArrayOfType(transformScalarType(type.ofType, state, objectProps));
  } else if (type instanceof require("graphql").GraphQLObjectType || type instanceof require("graphql").GraphQLUnionType || type instanceof require("graphql").GraphQLInterfaceType) {
    return objectProps;
  } else if (type instanceof require("graphql").GraphQLScalarType) {
    return transformGraphQLScalarType(type, state);
  } else if (type instanceof require("graphql").GraphQLEnumType) {
    return transformGraphQLEnumType(type, state);
  } else {
    throw new Error("Could not convert from GraphQL type ".concat(type.toString()));
  }
}

function transformGraphQLScalarType(type, state) {
  var customType = state.customScalars[type.name];

  switch (customType || type.name) {
    case 'ID':
    case 'String':
      return require("@babel/types").stringTypeAnnotation();

    case 'Float':
    case 'Int':
      return require("@babel/types").numberTypeAnnotation();

    case 'Boolean':
      return require("@babel/types").booleanTypeAnnotation();

    default:
      return customType == null ? require("@babel/types").anyTypeAnnotation() : require("@babel/types").genericTypeAnnotation(require("@babel/types").identifier(customType));
  }
}

function transformGraphQLEnumType(type, state) {
  state.usedEnums[type.name] = type;
  return require("@babel/types").genericTypeAnnotation(require("@babel/types").identifier(type.name));
}

function transformInputType(type, state) {
  if (type instanceof require("graphql").GraphQLNonNull) {
    return transformNonNullableInputType(type.ofType, state);
  } else {
    return require("@babel/types").nullableTypeAnnotation(transformNonNullableInputType(type, state));
  }
}

function transformNonNullableInputType(type, state) {
  if (type instanceof require("graphql").GraphQLList) {
    return require("./RelayFlowBabelFactories").readOnlyArrayOfType(transformInputType(type.ofType, state));
  } else if (type instanceof require("graphql").GraphQLScalarType) {
    return transformGraphQLScalarType(type, state);
  } else if (type instanceof require("graphql").GraphQLEnumType) {
    return transformGraphQLEnumType(type, state);
  } else if (type instanceof require("graphql").GraphQLInputObjectType) {
    var typeIdentifier = getInputObjectTypeIdentifier(type);

    if (state.generatedInputObjectTypes[typeIdentifier]) {
      return require("@babel/types").genericTypeAnnotation(require("@babel/types").identifier(typeIdentifier));
    }

    state.generatedInputObjectTypes[typeIdentifier] = 'pending';
    var fields = type.getFields();
    var props = Object.keys(fields).map(function (key) {
      return fields[key];
    }).map(function (field) {
      var property = require("@babel/types").objectTypeProperty(require("@babel/types").identifier(field.name), transformInputType(field.type, state));

      if (state.optionalInputFields.indexOf(field.name) >= 0 || !(field.type instanceof require("graphql").GraphQLNonNull)) {
        property.optional = true;
      }

      return property;
    });
    state.generatedInputObjectTypes[typeIdentifier] = require("./RelayFlowBabelFactories").exactObjectTypeAnnotation(props);
    return require("@babel/types").genericTypeAnnotation(require("@babel/types").identifier(typeIdentifier));
  } else {
    throw new Error("Could not convert from GraphQL type ".concat(type.toString()));
  }
}

module.exports = {
  transformInputType: transformInputType,
  transformScalarType: transformScalarType
};
