// relay-compiler/lib/RelaySourceModuleParser.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *  strict-local
 * @format
 */
'use strict';

var _toConsumableArray2 = require("@babel/runtime/helpers/interopRequireDefault")(require("@babel/runtime/helpers/toConsumableArray"));

var parseGraphQL = require("./GraphQLCompilerProfiler").instrument(require("graphql").parse, 'GraphQL.parse');

var FIND_OPTIONS = {
  validateNames: true
};

module.exports = function (tagFinder) {
  var memoizedTagFinder = require("./RelayFindGraphQLTags").memoizedFind.bind(null, tagFinder);

  function parseFile(baseDir, file) {
    var result = parseFileWithSources(baseDir, file);

    if (result) {
      return result.document;
    }
  } // Throws an error if parsing the file fails


  function parseFileWithSources(baseDir, file) {
    var text = require("fs").readFileSync(require("path").join(baseDir, file.relPath), 'utf8');

    !(text.indexOf('graphql') >= 0) ? process.env.NODE_ENV !== "production" ? require("fbjs/lib/invariant")(false, 'RelaySourceModuleParser: Files should be filtered before passed to the ' + 'parser, got unfiltered file `%s`.', file.relPath) : require("fbjs/lib/invariant")(false) : void 0;
    var astDefinitions = [];
    var sources = [];
    memoizedTagFinder(text, baseDir, file, FIND_OPTIONS).forEach(function (template) {
      var source = new (require("graphql").Source)(template, file.relPath);
      var ast = parseGraphQL(source);
      !ast.definitions.length ? process.env.NODE_ENV !== "production" ? require("fbjs/lib/invariant")(false, 'RelaySourceModuleParser: Expected GraphQL text to contain at least one ' + 'definition (fragment, mutation, query, subscription), got `%s`.', template) : require("fbjs/lib/invariant")(false) : void 0;
      sources.push(source.body);
      astDefinitions.push.apply(astDefinitions, (0, _toConsumableArray2["default"])(ast.definitions));
    });
    return {
      document: {
        kind: 'Document',
        definitions: astDefinitions
      },
      sources: sources
    };
  }

  function getParser(baseDir) {
    return new (require("./ASTCache"))({
      baseDir: baseDir,
      parse: parseFile
    });
  }

  function getFileFilter(baseDir) {
    return function (file) {
      var text = require("fs").readFileSync(require("path").join(baseDir, file.relPath), 'utf8');

      return text.indexOf('graphql') >= 0;
    };
  }

  return {
    getParser: getParser,
    getFileFilter: getFileFilter,
    parseFile: parseFile,
    parseFileWithSources: parseFileWithSources
  };
};
