// react-relay/lib/RelayMockRenderer.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */
'use strict';

var _inheritsLoose2 = require("@babel/runtime/helpers/interopRequireDefault")(require("@babel/runtime/helpers/inheritsLoose"));

var _assertThisInitialized2 = require("@babel/runtime/helpers/interopRequireDefault")(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = require("@babel/runtime/helpers/interopRequireDefault")(require("@babel/runtime/helpers/defineProperty"));

/**
 * A helper for rendering RelayContainers with mock data, outside of a
 * RelayRootContainer/RelayRenderer. This is intended for use in unit tests or
 * component browser-style interfaces.
 *
 * Note: For unit tests, you may need to mock `ReactDOM` as follows:
 *
 * ```
 * jest.mock('ReactDOM', () => ({}));
 * ```
 *
 * Currently ReactDOM and ReactTestRenderer cannot both be loaded in the same
 * test, and Relay transitively includes ReactDOM under its default settings.
 */
// this should be deprecated
var RelayMockRenderer =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2["default"])(RelayMockRenderer, _React$Component);

  function RelayMockRenderer() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])((0, _assertThisInitialized2["default"])(_this)), "mockContext", {
      environment: new (require("./RelayEnvironment"))(),
      variables: {},
      route: {
        name: '$RelayMockRenderer',
        params: {},
        queries: {},
        useMockData: true
      },
      useFakeData: true
    });
    return _this;
  }

  var _proto = RelayMockRenderer.prototype;

  _proto.render = function render() {
    return require("react").createElement(require("./ReactRelayContext").Provider, {
      value: this.mockContext
    }, this.props.render());
  };

  return RelayMockRenderer;
}(require("react").Component);

module.exports = RelayMockRenderer;
