// react-relay/lib/RelayOperationDescriptor.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *  strict-local
 * @format
 */
'use strict';

/**
 * @public
 *
 * Implementation of `RelayCore#createOperationDescriptor()` defined in
 * `RelayEnvironmentTypes` for the classic core.
 */
function createOperationDescriptor(operation, variables) {
  var concreteFragment = require("./QueryBuilder").getFragment(operation.node);

  !concreteFragment ? process.env.NODE_ENV !== "production" ? require("fbjs/lib/invariant")(false, 'RelayOperationDescriptor: Expected a query, got %s `%s`.', operation.node.kind, operation.name) : require("fbjs/lib/invariant")(false) : void 0;

  var operationVariables = require("./RelayVariables").getOperationVariables(operation, variables);

  var fragment = {
    dataID: require("./RelayStoreConstants").ROOT_ID,
    node: concreteFragment,
    variables: operationVariables
  };
  return {
    fragment: fragment,
    node: operation,
    root: fragment,
    variables: operationVariables
  };
}

module.exports = {
  createOperationDescriptor: createOperationDescriptor
};
