// react-relay/lib/RelayMutationRequest.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *  strict-local
 * @format
 */
'use strict';

var _inheritsLoose2 = require("@babel/runtime/helpers/interopRequireDefault")(require("@babel/runtime/helpers/inheritsLoose"));

/**
 * @internal
 *
 * Instances of these are made available via `RelayNetworkLayer.sendMutation`.
 */
var RelayMutationRequest =
/*#__PURE__*/
function (_Deferred) {
  (0, _inheritsLoose2["default"])(RelayMutationRequest, _Deferred);

  function RelayMutationRequest(mutation, files) {
    var _this;

    _this = _Deferred.call(this) || this;
    _this._mutation = mutation;
    _this._printedQuery = null;
    _this._files = files;
    return _this;
  }
  /**
   * @public
   *
   * Gets a string name used to refer to this request for printing debug output.
   */


  var _proto = RelayMutationRequest.prototype;

  _proto.getDebugName = function getDebugName() {
    return this._mutation.getName();
  };
  /**
   * @public
   *
   * Gets an optional map from name to File objects.
   */


  _proto.getFiles = function getFiles() {
    return this._files;
  };
  /**
   * @public
   *
   * Gets the variables used by the mutation. These variables should be
   * serialized and sent in the GraphQL request.
   */


  _proto.getVariables = function getVariables() {
    return this._getPrintedQuery().variables;
  };
  /**
   * @public
   *
   * Gets a string representation of the GraphQL mutation.
   */


  _proto.getQueryString = function getQueryString() {
    return this._getPrintedQuery().text;
  };
  /**
   * @public
   * @unstable
   */


  _proto.getMutation = function getMutation() {
    return this._mutation;
  };
  /**
   * @private
   *
   * Returns the memoized printed query.
   */


  _proto._getPrintedQuery = function _getPrintedQuery() {
    if (!this._printedQuery) {
      this._printedQuery = require("./printRelayQuery")(this._mutation);
    }

    return this._printedQuery;
  };

  return RelayMutationRequest;
}(require("fbjs/lib/Deferred"));

module.exports = RelayMutationRequest;
